File: src/c/src/merge_sort.c
// IMPLEMENTAÇÃO MERGE SORT EM C
#include "merge_sort.h"

void merge(int *v, int inicio, int meio, int fim) {
    int tamanho = fim - inicio + 1;
    int *temp = (int *) malloc(tamanho * sizeof(int));
    int p1 = inicio;
    int p2 = meio + 1;
    int fim1 = 0, fim2 = 0;
    int i, j, k;
   
    temp = (int *) malloc(tamanho * sizeof(int));
    if (temp != NULL) {
        for (i = 0; i < tamanho; i++) {
            if (!fim1 && !fim2) {
                if (v[p1] < v[p2]) {
                    temp[i] = v[p1++];
                } else {
                    temp[i] = v[p2++];
                }
                if (p1 > meio) {
                    fim1 = 1;
                }
                if (p2 > fim) {
                    fim2 = 1;
                }
            } else {
                if (!fim1) {
                    temp[i] = v[p1++];
                } else {
                    temp[i] = v[p2++];
                }
            }
        }
        for (j = 0, k = inicio; j < tamanho; j++, k++) {
            v[k] = temp[j];
        }
    }

    free(temp);
}

void mergeSort(int *v, int inicio, int fim) {
    int meio;
    if (inicio < fim) {
        meio = (inicio + fim) / 2;
        mergeSort(v, inicio, meio);
        mergeSort(v, meio + 1, fim);
        merge(v, inicio, meio, fim);
    }
}

void definirArray(int n, int *v, char *nome) {
    FILE *arq = fopen(nome, "r");
    if (arq == NULL) {
        perror("Erro ao abrir o arquivo de entrada");
        exit(EXIT_FAILURE);
    }

    int i = 0;
    for (i = 0; i < n && fscanf(arq, "%d", &v[i]) == 1; i++);

    fclose(arq);
}

void salvarTempo(int n, double time_taken, char *nome, char *nome2) {
    FILE *arq2 = fopen(nome2, "a");
    if (arq2 == NULL) {
        perror("Erro ao abrir o arquivo de saída");
        exit(EXIT_FAILURE);
    }

    fprintf(arq2, "\nC,%d,%f,%s", n, time_taken, nome);
    fclose(arq2);
}

void run(int n, char *nome, char *nome2) {
    int *v = (int *) malloc(n * sizeof(int));
    if (v == NULL) {
        perror("Erro ao alocar memória");
        exit(EXIT_FAILURE);
    }

    definirArray(n, v, nome);

    clock_t t = clock();
    mergeSort(v, 0, n - 1);
    t = clock() - t;
    double time_taken = ((double) t) / CLOCKS_PER_SEC;

    salvarTempo(n, time_taken, nome, nome2);
    printf("\nTempo de execução: %f", time_taken);
    free(v);
}

File: src/javascript/src/MergeSort.js
// IMPLEMENTAÇÃO MERGE SORT EM JAVASCRIPT
const fs = require('fs');

class MergeSort {
    constructor(arqE, arqS) {
        this.array = [];
        this.arq = arqE;
        this.arq2 = arqS;
    }

    merge(v, inicio, meio, fim) {
        let temp = [];
        let p1 = inicio, p2 = meio + 1;

        while (p1 <= meio && p2 <= fim) {
            if (v[p1] <= v[p2]) {
                temp.push(v[p1++]);
            } else {
                temp.push(v[p2++]);
            }
        }

        while (p1 <= meio) {
            temp.push(v[p1++]);
        }

        while (p2 <= fim) {
            temp.push(v[p2++]);
        }

        for (let i = 0; i < temp.length; i++) {
            v[inicio + i] = temp[i];
        }
    }

    mergeSort(v, inicio, fim) {
        let meio;
        if (inicio < fim) {
            meio = Math.floor((inicio + fim) / 2);
            this.mergeSort(v, inicio, meio);
            this.mergeSort(v, meio + 1, fim);
            this.merge(v, inicio, meio, fim);
        }
    }

    definirArray(n) {
        return new Promise((resolve, reject) => {
            fs.readFile(this.arq, 'utf8', (err, data) => {
                if (err) {
                    return reject(`Erro ao ler o arquivo de entrada: ${err.message}`);
                }

                this.array = data.split(/\s+/).map(Number);
                if (this.array.length > n) {
                    this.array = this.array.slice(0, n);
                }
                resolve();
            });
        });
    }

    salvarTempo(n, timeTaken) {
        const data = `\nJavaScript,${n},${timeTaken},${this.arq}`;
        fs.appendFile(this.arq2, data, (err) => {
            if (err) {
                console.error(`Erro ao salvar o tempo de execução: ${err.message}`);
            }
        });
        console.log(`Tempo de execução: ${timeTaken}`);
    }

    async run(n) {
        try {
            await this.definirArray(n);

            const startTime = process.hrtime();
            this.mergeSort(this.array, 0, this.array.length - 1);
            const endTime = process.hrtime(startTime);
            const timeTaken = (endTime[0] + endTime[1] / 1e9).toFixed(10);

            this.salvarTempo(n, timeTaken);
        } catch (error) {
            console.error(error);
        }
    }
}

module.exports = MergeSort;
